processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java

```diff
@@ -359,9 +360,19 @@ public Object apply(final Result<SelectResultValue> input)
     }
 
     final Granularity granularity = query.getGranularity();
+    final Iterable<String> filteredPagingKeys = Iterables.filter(
+        paging.keySet(), new Predicate<String>()
+        {
+          @Override
+          public boolean apply(String input)
+          {
+            return input.startsWith(String.format("%s_", dataSource));
```

Comment:
This isn't robust to a union dataSource of `["foo", "foo_2000"]`. Instead, use something like `DataSegmentUtils.valueOf(dataSource, input) != null`.

Although, that won't work as-is, since currently that method throws an exception rather than returning null if there isn't a match. But it only has one caller, so I think it'd be fine for you to change it to null return in this patch (and update that one caller, & some javadoc for good measure).